[a11y] Support TalkBack reading by word, character, and paragraph

jonahwilliams opened the pull request:
The logic in talkback does not handle reading from the text portion of an accessibility node when checking if that node could support reading at different granularities. See: https://github.com/google/talkback/blob/9b1d5132b074174d02886faccf731e814d69363c/talkback/src/main/java/GranularityTraversal.java#L207

Fix this by always setting the content description if something is not a text field.

https://github.com/flutter/flutter/issues/13406

Fixes https://github.com/flutter/flutter/issues/13406

goderbauer commented on shell/platform/android/io/flutter/view/AccessibilityBridge.java:
```diff
@@ -756,13 +761,7 @@ public AccessibilityNodeInfo createAccessibilityNodeInfo(int virtualViewId) {
       result.setChecked(semanticsNode.hasFlag(Flag.IS_TOGGLED));
       result.setClassName("android.widget.Switch");
-      result.setContentDescription(semanticsNode.getValueLabelHint());
-    } else if (!semanticsNode.hasFlag(Flag.SCOPES_ROUTE)) {
```
Does this change now make SemanticsNodes that are scoping a route and have a label focusable?

jonahwilliams replied:
I'll update the condition so we don't apply it to those nodes, thanks for jogging my memory a bit on the context

goderbauer commented:
/cc @xster Do you want to review this one?

xster commented:
Thanks for the fix @jonahwilliams. 

We should add a test for this. The good news is that writing Java tests is a lot easier now than before. See https://github.com/flutter/engine/blob/master/shell/platform/android/test/README.md#L24. The tests and dependencies all build in ninja and you can use robolectric, mockito and all that good stuff to construction inject etc. 

Give it a try. Happy to work with you on it if needed.

jonahwilliams commented:
@xster I updated run_test to use a configured Java so I can point it at Android Studio's JRE. However now I'm getting:

```
Exception in thread "main" java.lang.UnsupportedClassVersionError: io/flutter/FlutterTestSuite has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```

But this is 1.8.0_212-release-1586-b4-5784211 , so it should work according to the readme?

xster commented:
Are you on a mac? If you just hard-call a specific java binary but don't set your JAVA_HOME, it's gonna do really weird things. What does your JAVA_HOME and /usr/libexec/java_home -V return? 

i.e. I think it's going to be a bit more involved to changing the java instance than just calling a specific binary. If documentation is lacking, adding more docs to let people set JAVA_HOME is probably the least scary option.

xster commented on shell/platform/android/test/io/flutter/FlutterTestSuite.java:
```diff
@@ -29,6 +29,7 @@
 import test.io.flutter.embedding.engine.FlutterShellArgsTest;
 import test.io.flutter.embedding.engine.PluginComponentTest;
 import test.io.flutter.embedding.engine.dart.DartExecutorTest;
+import io.flutter.view.AccessibilityBridgeTest;
```
I assume the formatter will complain about this

jonahwilliams replied:
perhaps ... I will format another day :)

xster commented on testing/run_tests.py:
```diff
@@ -279,16 +279,16 @@ def EnsureJavaTestsAreBuilt(android_out_dir):
   EXPECTED_VERSION = '1.8'
   # `java -version` is output to stderr. https://bugs.java.com/bugdatabase/view_bug.do?bug_id=4380614
-  version_output = subprocess.check_output(['java', '-version'], stderr=subprocess.STDOUT)
+  version_output = subprocess.check_output([java_path, '-version'], stderr=subprocess.STDOUT)
```
if you really want this, you'll probably need an env map too

xster replied:
for JAVA_HOME

jonahwilliams replied:
Ahh yeah, with JAVA_HOME set everything is working. Thanks!

jonahwilliams commented on shell/platform/android/io/flutter/view/AccessibilityBridge.java:
```diff
@@ -1660,7 +1659,7 @@ public void reset() {
   // Must match SemanticsFlag in semantics.dart
   // https://github.com/flutter/engine/blob/master/lib/ui/semantics.dart
-  private enum Flag {
+  public enum Flag {
```
I'm not sure of the implications

xster replied:
```suggestion
  /* Package */ enum Flag {
```

xster replied:
doesn't need to be public if you just want it in a test (which is in the same package and default visibility is fine)

jonahwilliams commented on shell/platform/android/test/io/flutter/view/AccessibilityBridgeTest.java:
```diff
@@ -0,0 +1,185 @@
+package io.flutter.view;
+
+import static org.junit.Assert.assertEquals;
```
If the formatter didn't like the file above, it'll hate this

jonahwilliams commented:
I rolled back the run_test.py changes since that wasn't going to be sufficient. Test files are now formatted correctly.

I added the necessary test code to simulate semantic updates, though I only added tests for the conditions I changed

jonahwilliams commented:
Friendly ping @xster

xster commented on shell/platform/android/test/io/flutter/view/AccessibilityBridgeTest.java:
```diff
@@ -0,0 +1,188 @@
+  }
+
+  /// The encoding for semantics is described in platform_view_android.cc
+  class TestSemanticsUpdate {
```
if we want this mapping to be maintained manually, can we add notes to each of accessibilitybridge.updatesemantics, platformviewandroid.updatesemantics and here to remind maintainers?

jonahwilliams replied:
Something like : "This logic is also used in the test fixture at path/to/test.java"?

xster replied:
Sure. Something like "if you change stuff here, don't forget to update the test file as well"

jonahwilliams replied:
I added a comment to platform_view_android.cc, since that defines the encoding format. Elsewhere it is only consumed

xster commented:
LGTM!
